src/flags.zig

```diff
@@ -0,0 +1,127 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+/// Format and print an error message followed by the usage string to stderr,
+/// then exit with an exit code of 1.
+pub fn fatal(comptime fmt_string: []const u8, args: anytype) noreturn {
+    const stderr = std.io.getStdErr().writer();
+    stderr.print("error: " ++ fmt_string ++ "\n", args) catch {};
+    std.os.exit(1);
+}
+
+pub fn parse_commands(args: *std.process.ArgIterator, comptime Commands: type) Commands {
+    assert(@typeInfo(Commands) == .Union);
+
+    var subcommand = args.next() orelse fatal("subcommand required", .{});
+    inline for (comptime std.meta.fields(Commands)) |field| {
+        comptime assert(std.mem.indexOf(u8, field.name, "_") == null);
+        if (std.mem.eql(u8, subcommand, field.name)) {
+            return @unionInit(Commands, field.name, parse_flags(args, field.field_type));
+        }
+    }
+    fatal("unknown subcommand: '{s}'", .{subcommand});
+}
+
+pub fn parse_flags(args: *std.process.ArgIterator, comptime Flags: type) Flags {
+    if (Flags == void) {
+        if (args.next()) |arg| {
+            fatal("unexpected argument: '{s}'", .{arg});
+        }
+        return {};
+    }
+
+    assert(@typeInfo(Flags) == .Struct);
+
+    var result: Flags = undefined;
+    var counts: FlagsCounts(Flags) = .{};
+
+    next_arg: while (args.next()) |arg| {
+        inline for (comptime std.meta.fields(Flags)) |field| {
+            const flag = comptime field_to_flag(field.name);
+            if (std.mem.startsWith(u8, arg, flag)) {
```

Comment:
Maybe additionally check that the next character after the flag is either nothing, `=`, or a space?

Right now:

```
$ zig run  src/copyhound.zig -- memcpy --bytes_except_this_is_a_different_flag
error: --bytes argument requires a value
```

which is kind of weird, and I think would be broken if one flag was a prefix of another.

Reply:
Actually I think `tigerbeetle/cli.zig` has this exact problem too! 

```zig
        } else if (mem.startsWith(u8, arg, "--cache-transfers")) {
            if (command != .start) fatal("--cache-transfers: supported only by 'start' command", .{});
            cache_transfers = parse_flag("--cache-transfers", arg);
        } else if (mem.startsWith(u8, arg, "--cache-transfers-posted")) {
            if (command != .start) fatal("--cache-transfers-posted: supported only by 'start' command", .{});
            cache_transfers_posted = parse_flag("--cache-transfers-posted", arg);
```

Reply:
>which is kind of weird, and I think would be broken if one flag was a prefix of another.

Yeah, this is intentional bug-for-bug compatibility with our `cli.zig` for the moment. 

I do want to say "let's do the right thing when parsing CLI arguments", but at this point I am not quite sure what that right thing is

Reply:
Ah, well, this actually hits the benchmarks, which have `--transfer-count` and `--transfer-count-per-second`. I was thinking about just banning prefix args, but seems better to handle this properly.